### DISCUSSION TITLE
Automatic trigger via udev

### DISCUSSION THREAD

**Discussion by Jip-Hop** [psy0rz/zfs_autobackup#123] [Show and tell]
Hi @psy0rz!

ZFS-autobackup is looking really good! I made [a project to trigger it automatically](https://github.com/Jip-Hop/udev-trigger-zfs-autobackup) when my backup disk is attached (specifically made for TrueNAS SCALE).

I was wondering if you could take a look at [backup.sh](https://github.com/Jip-Hop/udev-trigger-zfs-autobackup/blob/main/backup.sh).

I was getting this error: `# ! [Source] STDERR > cannot hold snapshot 'data@offsite1-20220409123946': tag already exists on this dataset`. Do you know why this happens? Should I use `--no-holds` to prevent it? Is that safe? I don't fully understand the consequences xD

Also I got the `cannot receive incremental stream: destination has been modified since most recent snapshot` error, so I now set the dataset to readonly after the backup. But I could also use `--rollback` instead, right? Does the `--rollback` option cause data to be sent again (and thus slow the backup down)?

P.S. looking forward to incorporate `zfs-autoverify` once it's released! 🙂

**Comment by psy0rz** [comment]
Very nice project, il'l look at the script right away. And i like the concept of automatic backups on disk insertion!

You can ignore that 'tag already exists' error: That just means the dataset already had a hold-tag because of a failed backup or something.

regarding readonly: You can keep the readonly flag on, its possible to still receive new data. (readonly is only against modification via the mount point)

--rollback  will rollback the data to the last snapshot. e.g. the changes you or something made to the dataset. this will not result in extra data needing to be send or anything. in the perfect world a rollback shouldn't have to do anything, since there shouldnt be changes after the last backup. (and you shouldn't normally get that 'destination modified' error)

When you got a working version i'll link to it from the wiki,

**Reply by Jip-Hop** [reply]
> When you got a working version i'll link to it from the wiki,

Would you mind [trying it out](https://github.com/Jip-Hop/udev-trigger-zfs-autobackup)? I think it should now also work on different systems (Proxmox for example, not just TrueNAS SCALE).

**Comment by psy0rz** [comment]
A few notes:

- Its perfectly ok to import https://github.com/psy0rz/zfs_autobackup/blob/master/zfs_autobackup/ZfsAutobackup.py and use it directly from python. Might be easier if a part already is written in python.
- I'm not sure why you would have 'tag already exists' errors. perhaps because you're testing and sending the same snapshots again? Just ignore it or use --no-holds to make it faster. (holds are just to prevent accidental common snapshot deletion by the admin)
- I see you already set readonly on which is nice.
- I would use if ! zpool list ${POOL} 2>/dev/null .... to warn user of unsafe removal. (In case the user turned canmount=noauto on) Also it zfs cant get corrupted by unsafe removal, but you'll just mis the lastest data. (and the backup isnt complete) Next time it fill resume and finish the backup.

Script looks good!

I'm hungy so i'm not very sharp. Will eat now :)

**Reply by Jip-Hop** [reply]
> Script looks good!

Thank you, and thanks for looking!

I will use `--no-holds` then. Should be safe if I completely leave the backup pool alone: not manually remove datasets or snapshots, leave that to Thinner. Right?

> I would use if ! zpool list ${POOL} 2>/dev/null

Nice, I'll try that! I wasn't too happy about checking with `mountpoint` anyway.

> Also it zfs cant get corrupted by unsafe removal

Good to know ZFS (and zfs_autobackup) is pretty resilient to (accidentally) pulling the drive during a backup.

> Will eat now :)

I hope you had a nice dinner!

**Reply by Jip-Hop** [reply]
Does it hurt to use `--encrypt` when transferring from encrypted -> encrypted? I'm doing non-encrypted -> encrypted (already have [SED](https://github.com/Jip-Hop/sedunlocksrv-pba/) on source). And what happens if using `--encrypt` if the destination dataset isn't encrypted? Is `--encrypt` a safe default for my script?

**Reply by psy0rz** [reply]
Its a safe default, but checkout https://github.com/psy0rz/zfs_autobackup/wiki/Encryption

**Reply by psy0rz** [reply]
yes --no-holds is safe in this case.

**Comment by psy0rz** [comment]
also: zfs-verify can be very resource intensive, espdvially for big volumes.

i would recommend against doing it every time. Use it more as a monthly tool or something like that.

**Reply by Jip-Hop** [reply]
Could a 'schedule' be a feature of `zfs-verify`? Like there is a schedule to keep snapshots with Thinner? Or would that be up to the user? 🙂

I'm currently [used to verifying my entire backup](https://github.com/Jip-Hop/synology-autorun#hyper-backup--integrity-check-with-email-notifications) (runs overnight, ready the next day). So for my use case it's fine, but I also don't have especially big volumes.

**Comment by psy0rz** [comment]
i think the zfs-autoverify schedule is up to the user by using regular tools. (like cron or your script)
